src/main/java/frc/robot/subsystems/drive/Drive.java

```diff
@@ -157,18 +160,53 @@ public void arcadeDrive(double throttle, double turnModifier){
     differentialDrive.arcadeDrive(throttle, turnModifier);
   }
 
+  /**
+   * Executes curvatureDrive with closed loop control. Needs to be called
+   * periodically by the command to update the closed loop. 
+   * 
+   * The closed loop works on the difference between the desired curvature
+   * and the measured curvature, using proportional gain only.
+   * 
+   * @param throttle - [-1.0, 1.0]. Positive is forward
+   * @param curvature - [-1.0, 1.0]. Positive is turning right
+   */
   public void curvatureDrive(double throttle, double curvature) {
-    if(throttle != 0.0 && curvature == 0.0 && !gyroDisabled) { //driving straight and no turn
-      if(this.desiredAngle == Integer.MAX_VALUE) { //means robot just started driving straight
-        this.desiredAngle = gyroscope.getGyroAngle(); 
-      }
-      curvature = this.getAngularError(desiredAngle) * P_VALUE; 
-      this.setLeftPower(throttle - curvature);
-      this.setRightPower(throttle + curvature);
-    }
-    else { // when robot isn't driving straight
-      this.desiredAngle = Integer.MAX_VALUE; //if turn is greater than 0 or if robot is still
-      differentialDrive.curvatureDrive(-throttle, curvature, (Math.abs(throttle) < QUICK_TURN_THROTTLE_DEADZONE));
+
+    if(Math.abs(throttle) < QUICK_TURN_THROTTLE_DEADZONE){
+      // If throttle is less than the deadzone, treat inputs as arcade
+      this.arcadeDrive(throttle, curvature);
+    }else{
+      // Otherwise, treat inputs as curvature
+
+      // Calculate ideal left and right speeds based on curvature
+      WheelSpeeds desiredSpeeds = DifferentialDrive.curvatureDriveIK(throttle, curvature, false);
+
+      // To execute closed loop and correct unwanted turning, we need to directly
+      // compare a desired quanity to the current quantity to get an error.
+      // The units on desired speeds are hard to relate to turning units directly,
+      // so we need a way to cancel them out. If we control the curvature rate instead (which
+      // is turn rate / linear rate), we can cancel out some of the units into degrees/meter,
+      // making comparison easier.
+
+      // Desired curvature rate (angular speed / linear speed)
+      double desiredTurnRate = (180 / Math.PI) * (desiredSpeeds.left - desiredSpeeds.right) / (Constants.TRACK_WIDTH_METERS / 2);
+      double desiredSpeed = Math.abs(desiredSpeeds.left + desiredSpeeds.right) / 2.0;
+      // account for the possiblity of dividing by zero by setting miniumum "desired speed" value
+      double desiredCurvature = desiredTurnRate / Math.max(desiredSpeed, QUICK_TURN_THROTTLE_DEADZONE);
```

Comment:
So curvature is just angular speed/linear speed?

Reply:
I don't think linear speed will ever be 0 because throttle will be > QUICK_TURN_THROTTLE_DEADZONE

Reply:
> So curvature is just angular speed/linear speed?

Any point on a curve has the same curvature as some circle, and doing math based on that circle is a lot easier. Curvature is defined as `curvature = 1 / radius`, with radius referring to that circle. This works out intuitively: larger circles = flatter "sides" = lower curvature, just like this equation.

In circular motion, `linear_speed = angular_frequency (radians) x radius`, so after accounting for unit conversions you can see `angular_speed / linear_speed = angular_speed / (angular_speed * radius) = 1 / radius`, which is indeed curvature.

Reply:
> I don't think linear speed will ever be 0 because throttle will be > QUICK_TURN_THROTTLE_DEADZONE. Is there a need to do `Math.max(desiredSpeed, QUICK_TURN_THROTTLE_DEADZONE)`?

Mostly just caution. If some weird math inside `curvatureDriveIK` pushed idealized velocity below stick value it could cause some noisy values at low speeds, but I think this exceedingly unlikely. Any assurance we don't divide by 0 is probably good enough.